app/controllers/tests_controller.rb

```diff
@@ -0,0 +1,151 @@
+class TestsController < ApplicationController
+  before_action :require_login
+  before_action :load_test, only: [:update, :show, :edit]
+  before_action :has_authorization_with_test?, only: [:show]
+
+  def create
+    @test = current_user.tests.build(test_params)
+    if @test.save
+      add_questions_to_test
+      redirect_to edit_test_path @test
+    else
+      flash.now[:danger] = t "tests.errors.questions_fail"
+      @subject = Subject.find_by id: params[:subject_id]
+      render "subjects/show"
+    end
+  end
+
+  def index
+    redirect_back fallback_location: request.referer
+  end
+
+  def show; end
+
+  def update
+    test_questions_ids = @test.test_questions
+                              .order_asc_question_ids.pluck(:id)
+    update_params = params.dig(:test, :question)&.values
+    detail_answers = []
+
+    return unless update_params
+
+    answer_ids = update_params.map{|i| i[:id]}
+    update_params.each_with_index do |update_param, index|
+      next if check_invalid_answer(update_param[:id])
+
+      add_answer_to_question(test_questions_ids, update_param[:id],
+                             detail_answers, index)
+    end
+    ActiveRecord::Base.transaction do
+      create_relations(detail_answers)
+      update_test(check_score(answer_ids))
+    end
+  end
+
+  def edit; end
+
+  private
+
+  def check_invalid_answer answer
+    return true if answer == "" ||
+                   (answer.is_a?(Array) && answer.size == 1)
+  end
+
+  def update_test score
+    rate = score * 1.0 / @test.subject.question_amount * 100
+    if rate > @test.subject.pass_score
+      @test.update({status: Test.statuses[:passed], score: score})
+    else
+      @test.update({status: Test.statuses[:failed], score: score})
+    end
```

Comment:
```suggestion

    @test.update! status: (rate > @test.subject.pass_score ? Test.statuses[:passed] : Test.statuses[:fail]), score: score
```

ri dc ko

với lại nếu có transaction thì em phải dùng các method như update! thì mới có thể rollback data dc